core/vm/contracts_write.go

```diff
@@ -0,0 +1,161 @@
+	RunWriteCapable(evm *EVM, contract *Contract, input []byte) ([]byte, error)
+}
+
+// RunWriteCapablePrecompiledContract runs and evaluates the output of a write capable precompiled contract.
+func RunWriteCapablePrecompiledContract(p WriteCapablePrecompiledContract, evm *EVM, contract *Contract, input []byte, readOnly bool) ([]byte, error) {
+	// immediately error out if readOnly
+	if readOnly {
+		return nil, errWriteProtection
+	}
+	gas := p.RequiredGas(evm, input)
+	if !contract.UseGas(gas) {
+		return nil, ErrOutOfGas
+	}
+	return p.RunWriteCapable(evm, contract, input)
+}
+
+type stakingPrecompile struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *stakingPrecompile) RequiredGas(evm *EVM, input []byte) uint64 {
+	// If you check the tests, they contain a bunch of successful transactions
+	// which are of the type Delegate / Undelegate / CollectRewards
+	// these consume anywhere between 22-25k of gas, which is in line with
+	// what would happen for a non-EVM staking tx
+	// so no additional gas needs to be consumed here
+	return 0
+	// nb: failed transactions result in consumption of
+	// all the gas that was passed, as per core/vm/evm.go
+	// if err != ErrExecutionReverted {
+	//	contract.UseGas(contract.Gas)
+	// }
+	// which is why the solidity lib passes 0 as first param
+	// instead of the original gas(). if you do so, there will
+	// be no gas left for subsequent calls to the precompile
+	// (if you make multiple precompile calls in the same tx)
+}
+
+// RunWriteCapable runs the actual contract (that is it performs the staking)
+func (c *stakingPrecompile) RunWriteCapable(evm *EVM, contract *Contract, input []byte) ([]byte, error) {
+	// checks include input, shard, method
+	// readOnly has already been checked by RunWriteCapablePrecompiledContract
+	// first 32 for size and next 4 for method, at a minimum
+	if len(input) < 36 {
+		return nil, errors.New("Input is malformed")
+	}
+	if evm.Context.ShardID != shard.BeaconChainShardID {
+		return nil, errors.New("Staking not supported on this shard")
+		// we are not shard 0, so no processing
+		// but do not fail silently
+	}
+	input = input[32:] // drop the initial start of word information
+	method, err := staking.ParseStakingMethod(input)
+	if err != nil {
+		return nil, stakingTypes.ErrInvalidStakingKind
+	}
+	input = input[4:] // drop the method selector
+	// store passed information in map
+	args := map[string]interface{}{}
+	if err = method.Inputs.UnpackIntoMap(args, input); err != nil {
+		return nil, err
+	}
+	switch method.Name {
+	case "Delegate":
+		{
+			// a contract should only delegate its own balance - nobody else's
+			address, err := staking.ValidateContractAddress(contract.Caller(), args, "delegatorAddress")
```

Comment:
So if the user send a txn that directly calls the precompile contract address, then the staking will happen on the users EOA address right? Just as what's happening right now with Non-evm staking.

Reply:
Yes, the latest push (https://github.com/harmony-one/harmony/pull/3906/commits/c0765818add7e2bd8d73a5489a40841d549096e7) allows for this to happen. The associated change has been made to the Solidity lib.

Reply:
What happens if the precompile is called through another contract? The user have to trust that contract right? The contract is basically a delegate to the user.

Reply:
Sorry, missed this comment. 

EOAs can either
 - directly call the precompile, as demonstrated in [`test_eoa_access`](https://github.com/MaxMustermann2/harmony-staking-precompiles/blob/eca28448813ef90507bdda0e24909728233d49b1/tests/test_staking_contract.py#L778), or
 - use a contract which uses assembly `delegatecall`, as used in [`TrySubsidizeContract.sol`](https://github.com/MaxMustermann2/harmony-staking-precompiles/blob/eca28448813ef90507bdda0e24909728233d49b1/contracts/example/TrySubsidizeContract.sol#L29). This contract needs to be trusted by the user, yes. I have demonstrated its usage in [`test_eoa_subsidize_success`](https://github.com/MaxMustermann2/harmony-staking-precompiles/blob/eca28448813ef90507bdda0e24909728233d49b1/tests/test_staking_contract.py#L830), which also demonstrates that such a contract, which calls the precompile on behalf of the EOA, cannot be used to subsidize the gas for delegate transactions. A smart contract can also use this contract to call the precompile on its behalf.